src/runtime_src/core/edge/drm/zocl/edge/zocl_hwctx.c

```diff
@@ -0,0 +1,460 @@
+
+out:
+    mutex_unlock(&client->lock);
+    return ret;
+}
+
+int zocl_close_cu_ctx(struct drm_zocl_dev *zdev, struct drm_zocl_close_cu_ctx *drm_cu_ctx, struct kds_client *client)
+{
+    struct kds_client_hw_ctx *kds_hw_ctx = NULL;
+    struct kds_client_cu_ctx *kds_cu_ctx = NULL;
+    struct kds_client_cu_info kds_cu_info = {};
+    int ret = 0;
+
+    mutex_lock(&client->lock);
+
+    kds_hw_ctx = kds_get_hw_ctx_by_id(client, drm_cu_ctx->hw_context);
+    if (!kds_hw_ctx) {
+        DRM_ERROR("%s: No valid hw context is open", __func__);
+        ret = -EINVAL;
+        goto out;
+    }
+
+    zocl_close_cu_ctx_to_info(drm_cu_ctx, &kds_cu_info);
+
+    kds_cu_ctx = kds_get_cu_hw_ctx(client, kds_hw_ctx, &kds_cu_info);
+    if (!kds_cu_ctx) {
+        DRM_ERROR("%s: No cu context is open", __func__);
+        ret = -EINVAL;
+        goto out;
+    }
+
+    ret = kds_del_context(&zdev->kds, client, kds_cu_ctx);
+    if (ret)
+        goto out;
+
+    ret = kds_free_cu_ctx(client, kds_cu_ctx);
+
+out:
+    mutex_unlock(&client->lock);
+    return ret;
+}
+
+/**
+ * Callback function for async dma operation. This will also clean the
+ * command memory.
+ *
+ * @arg:    kds command pointer
+ * @ret:    return value of the dma operation.
+ */
+static void zocl_kds_dma_complete(void *arg, int ret)
+{
+    struct kds_command *xcmd = (struct kds_command *)arg;
+    zocl_dma_handle_t *dma_handle = (zocl_dma_handle_t *)xcmd->priv;
+
+    xcmd->status = KDS_COMPLETED;
+    if (ret)
+        xcmd->status = KDS_ERROR;
+    xcmd->cb.notify_host(xcmd, xcmd->status);
+    xcmd->cb.free(xcmd);
+
+    kfree(dma_handle);
+}
+
+/**
+ * Copy the user space command to kds command. Also register the callback
+ * function for the DMA operation.
+ *
+ * @zdev:   zocl device structure
+ * @flip:   DRM file private data
+ * @ecmd:   ERT command structure
+ * @xcmd:   KDS command structure
+ *
+ * @return      0 on success, Error code on failure.
+ */
+static int copybo_ecmd2xcmd(struct drm_zocl_dev *zdev, struct drm_file *filp,
+			    struct ert_start_copybo_cmd *ecmd,
+			    struct kds_command *xcmd)
+{
+    struct drm_device *dev = zdev->ddev;
+    zocl_dma_handle_t *dma_handle;
+    struct drm_zocl_copy_bo args = {
+        .dst_handle = ecmd->dst_bo_hdl,
+        .src_handle = ecmd->src_bo_hdl,
+        .size = ert_copybo_size(ecmd),
+        .dst_offset = ert_copybo_dst_offset(ecmd),
+        .src_offset = ert_copybo_src_offset(ecmd),
+    };
+    int ret = 0;
+
+    dma_handle = kmalloc(sizeof(zocl_dma_handle_t), GFP_KERNEL);
+    if (!dma_handle)
+        return -ENOMEM;
+
+    memset(dma_handle, 0, sizeof(zocl_dma_handle_t));
+
+    ret = zocl_dma_channel_instance(dma_handle, zdev);
+    if (ret)
+        return ret;
+
+    /* We must set up callback for async dma operations. */
+    dma_handle->dma_func = zocl_kds_dma_complete;
+    dma_handle->dma_arg = xcmd;
+    xcmd->priv = dma_handle;
+
+    ret = zocl_copy_bo_async(dev, filp, dma_handle, &args);
```

Comment:
```suggestion
    return zocl_copy_bo_async(dev, filp, dma_handle, &args);
```

Reply:
hi Saif, made the change on the latest push. Thanks